refactor endpoint tag management

chiptus opened the pull request:
Closes #3627

chiptus commented:
fix:
- when navigation to the home page as non-admin, tags are failing to load (auth problem) and endpoint are seen with no tags (although they have tags)
- endpoints list on home page is visible before tags are loaded, so it appears for 1s that the endpoints are missing their tags

chiptus commented:
@deviantony do you think we can allow non-admin users to get the tags list?

chiptus commented:
when deleting a tag and going to an endpoint that used to be associated with it, it is still associated with. before this change it wasn't a problem since tag was only a string, but now the tag-selector is throwing errors

chiptus commented:
> when deleting a tag and going to an endpoint that used to be associated with it, it is still associated with. before this change it wasn't a problem since tag was only a string, but now the tag-selector is throwing errors

fixed by changing the delete tag handler

deviantony commented on api/bolt/migrator/migrator.go:
```diff
@@ -301,5 +305,12 @@ func (m *Migrator) Migrate() error {
 		}
 	}
 
+	if m.currentDBVersion < 23 {
```
Add the following comment on top of this condition:

```golang
// Portainer 1.24.0-dev
```

chiptus replied:
done

deviantony commented on api/bolt/migrator/migrate_dbversion22.go:
```diff
@@ -0,0 +1,49 @@
+func (m *Migrator) updateEndointsAndEndpointsGroupsToDBVersion23() error {
+
+	tags, err := m.tagService.Tags()
+
```
Remove unnecessary extra line.

deviantony commented on api/bolt/migrator/migrate_dbversion22.go:
```diff
@@ -0,0 +1,49 @@
+	for _, endpoint := range endpoints {
+		endpointTags := make([]portainer.TagID, 0)
+		for _, tagName := range endpoint.Tags {
+			tagID := tagsNameMap[tagName]
```
what happens if the tag is not in the map? I think you should use the map check pattern here:

```
tagID, ok := tagsNameMap[tagName]
if ok {
// tagName found
}
```

deviantony commented on api/bolt/migrator/migrate_dbversion22.go:
```diff
@@ -0,0 +1,49 @@
+			endpointTags = append(endpointTags, tagID)
+		}
+		endpoint.TagIDs = endpointTags
+		m.endpointService.UpdateEndpoint(endpoint.ID, &endpoint)
```
Unhandled error.

deviantony commented on api/bolt/migrator/migrate_dbversion22.go:
```diff
@@ -0,0 +1,49 @@
+	for _, endpointGroup := range endpointGroups {
+		endpointGroupTags := make([]portainer.TagID, 0)
+		for _, tagName := range endpointGroup.Tags {
+			tagID := tagsNameMap[tagName]
```
what happens if the tag is not in the map? I think you should use the map check pattern here:

```
tagID, ok := tagsNameMap[tagName]
if ok {
// tagName found
}
```

deviantony commented on api/bolt/migrator/migrate_dbversion22.go:
```diff
@@ -0,0 +1,49 @@
+			endpointGroupTags = append(endpointGroupTags, tagID)
+		}
+		endpointGroup.TagIDs = endpointGroupTags
+		m.endpointGroupService.UpdateEndpointGroup(endpointGroup.ID, &endpointGroup)
```
Unhandled error.

deviantony commented on api/bolt/migrator/migrator.go:
```diff
@@ -37,6 +38,7 @@ type (
 		userService            *user.Service
 		versionService         *version.Service
 		fileService            portainer.FileService
+		tagService             *tag.Service
```
Relocate under `userService` to follow alphabetic order in this file. All the services of this package should follow this rule, services from outside this package can be put at the end of object definition (like the `portainer.FileService`).

deviantony commented on api/bolt/migrator/migrator.go:
```diff
@@ -57,6 +59,7 @@ type (
 		UserService            *user.Service
 		VersionService         *version.Service
 		FileService            portainer.FileService
+		TagService             *tag.Service
```
See my comment above about presenting properties using the alphabetical order.

deviantony commented on api/bolt/migrator/migrator.go:
```diff
@@ -79,6 +82,7 @@ func NewMigrator(parameters *Parameters) *Migrator {
 		userService:            parameters.UserService,
 		versionService:         parameters.VersionService,
 		fileService:            parameters.FileService,
+		tagService:             parameters.TagService,
```
See my comment above about presenting properties using the alphabetical order.

deviantony commented on api/bolt/tag/tag.go:
```diff
@@ -52,6 +52,19 @@ func (service *Service) Tags() ([]portainer.Tag, error) {
 	return tags, err
 }
 
+// Tag returns an tag by ID.
```
```suggestion
// Tag returns a tag by ID.
```

deviantony commented on api/http/handler/endpoints/endpoint_list.go:
```diff
@@ -51,7 +51,15 @@ func (handler *Handler) endpointList(w http.ResponseWriter, r *http.Request) *ht
+		if err != nil {
+			return &httperror.HandlerError{http.StatusInternalServerError, "Unable to retrieve tags from the database", err}
+		}
+		tagsMap := make(map[portainer.TagID]portainer.Tag)
```
What about using a `map[portainer.TagID]string` instead? No need to reference the entire tag since we're only using the name for search/comparison.

chiptus replied:
:+1:

deviantony commented on api/http/handler/tags/tag_delete.go:
```diff
@@ -15,6 +15,55 @@ func (handler *Handler) tagDelete(w http.ResponseWriter, r *http.Request) *httpe
+	}
+
+	for _, endpoint := range endpoints {
+		endpointTags := []portainer.TagID{}
```
I would introduce a new function to delete an element based on index (search and break instead of looping through all tags all the time).

That function could leverage the snippet described in https://github.com/golang/go/wiki/SliceTricks#delete-without-preserving-order

```golang
a[i] = a[len(a)-1]
a[len(a)-1] = nil
a = a[:len(a)-1]
```

Then I would search for the index of the tag to remove, if present remove and update.

Something like this is a lot more readable I think:

```golang
	for _, endpoint := range endpoints {
		tagIdx := 0
		for idx, tag := range endpoint.TagIDs {
			if tag == tagID {
				tagIdx = idx
				break
			}
		}

		if tagIdx != 0 {
			// tag index found, remove and update
		}
	}

```

deviantony commented on api/http/handler/tags/tag_delete.go:
```diff
@@ -15,6 +15,55 @@ func (handler *Handler) tagDelete(w http.ResponseWriter, r *http.Request) *httpe
+		return &httperror.HandlerError{http.StatusInternalServerError, "Unable to retrieve endpoints from the database", err}
+	}
+
+	for _, endpointGroup := range endpointGroups {
```
Same as for endpoints above.

deviantony commented on app/portainer/components/endpoint-list/endpoint-item/endpoint-item-controller.js:
```diff
@@ -1,12 +1,39 @@
-angular.module('portainer.app').controller('EndpointItemController', [
-  function EndpointItemController() {
-    var ctrl = this;
+angular.module('portainer.app').controller('EndpointItemController', function EndpointItemController() {
```
We'll want to talk to keep consistency inside the project for code style. 

These cosmetic changes actually makes the review process harder and break the consistency of the app.

deviantony commented on app/portainer/components/endpoint-list/endpoint-item/endpoint-item-controller.js:
```diff
@@ -1,12 +1,39 @@
+    if (!this.tags) {
+      return 'Loading tags...'
+    }
+    return this.model.TagIds.map(tagId => {
```
I would advocate against returning a function for clarity/readility.

Prepare the variable, assign it to your function and then returns the function. Shorter code do not make it easier to maintain. Clear code will.

chiptus replied:
not sure what you mean, something like this?
```
function findTagName(tag) {
  ...
}
return this.model.TagIds.map(findTagName);
```

chiptus replied:
the advantage of using arrow function is that the binding of `this` doesn't change. with this solution, we would need to rebind `this`

chiptus replied:
please have a look at my fix

deviantony commented on app/portainer/components/endpoint-list/endpoint-item/endpoint-item-controller.js:
```diff
@@ -1,12 +1,39 @@
+      }
+      return tag.Name;
+    })
+      .filter(Boolean)
```
Check indentation.

chiptus commented:
> fix:
> 
> * when navigation to the home page as non-admin, tags are failing to load (auth problem) and endpoint are seen with no tags (although they have tags)
fixed by changing the tag_list handler from `AdminAccess` to `AuthenticatedAccess`
> * endpoints list on home page is visible before tags are loaded, so it appears for 1s that the endpoints are missing their tags
fixed by showing `Loading tags...` label instead of the tags label

xAt0mZ commented on app/portainer/views/home/homeController.js:
```diff
@@ -1,6 +1,6 @@
-  .controller('HomeController', ['$q', '$scope', '$state', '$interval', 'Authentication', 'EndpointService', 'EndpointHelper', 'GroupService', 'Notifications', 'EndpointProvider', 'StateManager', 'LegacyExtensionManager', 'ModalService', 'MotdService', 'SystemService',
-    function($q, $scope, $state, $interval, Authentication, EndpointService, EndpointHelper, GroupService, Notifications, EndpointProvider, StateManager, LegacyExtensionManager, ModalService, MotdService, SystemService) {
+  .controller('HomeController', 
+    function($q, $scope, $state, TagService, Authentication, EndpointService, EndpointHelper, GroupService, Notifications, EndpointProvider, StateManager, LegacyExtensionManager, ModalService, MotdService, SystemService) {
```
**Simple comment**
Nice use of `babel-plugin-angularjs-annotate` auto detection !

chiptus replied:
yes, I realized that adding to both array and arguments would look like a lot of changes anyway :D

chiptus replied:
thanks

xAt0mZ commented on app/portainer/components/multi-endpoint-selector/multiEndpointSelectorController.js:
```diff
@@ -1,37 +1,42 @@
-});
+function filterEmptyGroups(groups, endpoints) {
+  return _.filter(groups, group => _.some(endpoints, endpoint => endpoint.GroupId == group.Id));
+}
```
Don't extract this filter in an external function, write `_.filter` ... in `OnInit` as it's the only usage occurrence in the project

+ add parenthesis around params of anonymous functions, even with a single param

xAt0mZ commented on app/portainer/components/tag-selector/tagSelectorController.js:
```diff
@@ -1,32 +1,34 @@
-    }
+angular.module('portainer.app').controller('TagSelectorController', function() {
+  this.$onInit = function() {
+    this.state.selectedTags = this.model.map(id => this.tags.find(t => t.Id === id));
```
Use lodash

xAt0mZ commented on app/portainer/components/tag-selector/tagSelectorController.js:
```diff
@@ -1,32 +1,34 @@
+      this.model.splice(index, 1);
+      this.state.selectedTags.splice(index, 1);
     }
   };
```
Use lodash `_.remove()`

xAt0mZ commented on app/portainer/components/tag-selector/tagSelectorController.js:
```diff
@@ -1,32 +1,34 @@
+    if (!this.model) {
+      return true;
+    }
+    return !this.model.includes($item.Id);
```
use lodash

xAt0mZ commented:
Some small changes, but we are nearly good to go !

deviantony commented:
I'll request another functional validation since the reviews introduced quite a few changes.

deviantony commented:
@xAt0mZ assuming this is looking ok to unlock other tasks, no need for another review.
